File: IUHCODER/Một số bài quy hoạch động kinh điển/LIS 1.py
from math import *

n = int(input())
a = list(map(int, input().split()))

dp = [1] * (n + 1)
ans = 0

for i in range(1, n):
    for j in range(0, i):
        if a[j] < a[i]:
            dp[i] = max(dp[i], dp[j] + 1)
    ans = max(ans, dp[i])
    
print(ans)

File: IUHCODER/Một số bài quy hoạch động kinh điển/LIS 1.cpp
#include <bits/stdc++.h>
using namespace std;

#define endl '\n'
#define int long long
#define inp freopen("file.inp", "r", stdin)
#define out freopen("file.out", "w", stdout)
#define TIME 1.0*clock()/CLOCKS_PER_SEC
#define fastIO ios_base::sync_with_stdio(0); cin.tie(0)

const int MAXN = 1e6 + 5;
const int MOD = 1e9 + 7;

int n, ans;
int arr[MAXN];
int dp[MAXN];

signed main() {
    fastIO;
    cin >> n;
    for (int i = 1; i <= n; i++) 
        cin >> arr[i];
    
    for (int i = 1; i <= n; i++) {
        dp[i] = 1;
        for (int j = 1; j < i; j++)
            if (arr[j] < arr[i]) 
                dp[i] = max(dp[i], dp[j] + 1);
        ans = max(ans, dp[i]);
    }
    
    cout << ans;
    return 0;
}